File: back-end/ShopHangTet/Controllers/CartController.cs
using Microsoft.AspNetCore.Mvc;
using ShopHangTet.DTOs;
using ShopHangTet.Services;
using System.Security.Claims;

namespace ShopHangTet.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class CartController : ControllerBase
    {
        private readonly ICartService _cartService;

        public CartController(ICartService cartService)
        {
            _cartService = cartService;
        }

        private (string? UserId, string? SessionId) GetUserOrSession()
        {
            var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value
                      ?? User.FindFirst("Id")?.Value
                      ?? User.FindFirst("id")?.Value
                      ?? User.FindFirst(System.IdentityModel.Tokens.Jwt.JwtRegisteredClaimNames.Sub)?.Value;

            var sessionId = Request.Headers["X-Session-Id"].FirstOrDefault();

            if (!string.IsNullOrWhiteSpace(userId))
            {
                sessionId = null;
            }

            return (userId, sessionId);
        }

        private bool IsValidIdentity(string? userId, string? sessionId)
        {
            return !string.IsNullOrWhiteSpace(userId) || !string.IsNullOrWhiteSpace(sessionId);
        }

        [HttpGet]
        public async Task<IActionResult> GetCart()
        {
            var (userId, sessionId) = GetUserOrSession();

            if (!IsValidIdentity(userId, sessionId))
                return BadRequest(ApiResponse<CartDto>.ErrorResult("Cần Token hoặc X-Session-Id"));

            var result = await _cartService.GetCartAsync(userId, sessionId);

            return result.Success ? Ok(result) : BadRequest(result);
        }

        [HttpPost("add")]
        public async Task<IActionResult> AddToCart([FromBody] AddToCartDto dto)
        {
            var (userId, sessionId) = GetUserOrSession();

            if (!IsValidIdentity(userId, sessionId))
                return BadRequest(ApiResponse<CartDto>.ErrorResult("Cần Token hoặc X-Session-Id"));

            var result = await _cartService.AddToCartAsync(userId, sessionId, dto);

            return result.Success ? Ok(result) : BadRequest(result);
        }

        [HttpPost("add-batch")]
        public async Task<IActionResult> AddToCartBatch([FromBody] AddToCartBatchDto dto)
        {
            var (userId, sessionId) = GetUserOrSession();

            if (!IsValidIdentity(userId, sessionId))
                return BadRequest(ApiResponse<CartDto>.ErrorResult("Cần Token hoặc X-Session-Id"));

            var result = await _cartService.AddToCartBatchAsync(userId, sessionId, dto);

            return result.Success ? Ok(result) : BadRequest(result);
        }

        [HttpPut("update/{itemId}")]
        public async Task<IActionResult> UpdateCartItem(string itemId, [FromBody] UpdateCartItemDto dto)
        {
            var (userId, sessionId) = GetUserOrSession();

            if (!IsValidIdentity(userId, sessionId))
                return BadRequest(ApiResponse<CartDto>.ErrorResult("Cần Token hoặc X-Session-Id"));

            var result = await _cartService.UpdateCartItemAsync(userId, sessionId, itemId, dto);

            return result.Success ? Ok(result) : BadRequest(result);
        }

        [HttpDelete("remove/{itemId}")]
        public async Task<IActionResult> RemoveFromCart(string itemId)
        {
            var (userId, sessionId) = GetUserOrSession();

            if (!IsValidIdentity(userId, sessionId))
                return BadRequest(ApiResponse<CartDto>.ErrorResult("Cần Token hoặc X-Session-Id"));

            var result = await _cartService.RemoveFromCartItemAsync(userId, sessionId, itemId);

            return result.Success ? Ok(result) : BadRequest(result);
        }

        [HttpDelete("clear")]
        public async Task<IActionResult> ClearCart()
        {
            var (userId, sessionId) = GetUserOrSession();

            if (!IsValidIdentity(userId, sessionId))
                return BadRequest(ApiResponse<CartDto>.ErrorResult("Cần Token hoặc X-Session-Id"));

            var result = await _cartService.ClearCartAsync(userId, sessionId);

            return result.Success ? Ok(result) : BadRequest(result);
        }
    }
}

File: back-end-nodejs/src/controllers/cartController.js
import { Router } from 'express';
import { ApiResponse } from '../dtos/index.js';

/**
 * Cart Controller - Tương đương CartController.cs
 * Routes: /api/cart
 */
export function createCartRouter(cartService) {
    const router = Router();

    /**
     * Lấy userId từ JWT token hoặc sessionId từ header
     */
    function getUserOrSession(req) {
        const userId = req.user?.sub || req.user?.id || null;
        const sessionId = req.headers['x-session-id'] || null;
        return { userId, sessionId };
    }

    function isValidIdentity(userId, sessionId) {
        return !!(userId || sessionId);
    }

    // ========== GET /api/cart ==========
    router.get('/', async (req, res) => {
        try {
            const { userId, sessionId } = getUserOrSession(req);

            if (!isValidIdentity(userId, sessionId)) {
                return res.status(400).json(ApiResponse.error('Cần Token hoặc X-Session-Id'));
            }

            const result = await cartService.getCart(userId, sessionId);
            return result.Success ? res.status(200).json(result) : res.status(400).json(result);
        } catch (error) {
            console.error('GetCart error:', error);
            return res.status(500).json(ApiResponse.error('Internal server error'));
        }
    });

    // ========== POST /api/cart/add ==========
    router.post('/add', async (req, res) => {
        try {
            const { userId, sessionId } = getUserOrSession(req);

            if (!isValidIdentity(userId, sessionId)) {
                return res.status(400).json(ApiResponse.error('Cần Token hoặc X-Session-Id'));
            }

            const result = await cartService.addToCart(userId, sessionId, req.body);
            return result.Success ? res.status(200).json(result) : res.status(400).json(result);
        } catch (error) {
            console.error('AddToCart error:', error);
            return res.status(500).json(ApiResponse.error('Internal server error'));
        }
    });

    // ========== PUT /api/cart/update/:itemId ==========
    router.put('/update/:itemId', async (req, res) => {
        try {
            const { userId, sessionId } = getUserOrSession(req);

            if (!isValidIdentity(userId, sessionId)) {
                return res.status(400).json(ApiResponse.error('Cần Token hoặc X-Session-Id'));
            }

            const result = await cartService.updateCartItem(userId, sessionId, req.params.itemId, req.body);
            return result.Success ? res.status(200).json(result) : res.status(400).json(result);
        } catch (error) {
            console.error('UpdateCartItem error:', error);
            return res.status(500).json(ApiResponse.error('Internal server error'));
        }
    });

    // ========== DELETE /api/cart/remove/:itemId ==========
    router.delete('/remove/:itemId', async (req, res) => {
        try {
            const { userId, sessionId } = getUserOrSession(req);

            if (!isValidIdentity(userId, sessionId)) {
                return res.status(400).json(ApiResponse.error('Cần Token hoặc X-Session-Id'));
            }

            const result = await cartService.removeFromCart(userId, sessionId, req.params.itemId);
            return result.Success ? res.status(200).json(result) : res.status(400).json(result);
        } catch (error) {
            console.error('RemoveFromCart error:', error);
            return res.status(500).json(ApiResponse.error('Internal server error'));
        }
    });

    // ========== DELETE /api/cart/clear ==========
    router.delete('/clear', async (req, res) => {
        try {
            const { userId, sessionId } = getUserOrSession(req);

            if (!isValidIdentity(userId, sessionId)) {
                return res.status(400).json(ApiResponse.error('Cần Token hoặc X-Session-Id'));
            }

            const result = await cartService.clearCart(userId, sessionId);
            return result.Success ? res.status(200).json(result) : res.status(400).json(result);
        } catch (error) {
            console.error('ClearCart error:', error);
            return res.status(500).json(ApiResponse.error('Internal server error'));
        }
    });

    return router;
}
